bokeh/application/handlers/document_lifecycle.py

```diff
@@ -62,8 +62,23 @@ def _on_session_destroyed(session_context):
     '''
     Calls any on_session_destroyed callbacks defined on the Document
     '''
-    for callback in session_context._document._session_destroyed_callbacks:
-        callback(session_context)
+    callbacks = list(session_context._document._session_destroyed_callbacks)
+    session_context._document._session_destroyed_callbacks = {}
```

Comment:
Let's try to define some minimal API so that it can be put under test

Reply:
I've amended the existing test to check they are in fact cleared, do you want me to add ``Document.clear_on_session_destroyed`` or similar?

Reply:
how about just a (Python) property getter/setter for `_session_destroyed_callbacks` ?

Reply:
Sure.